server/src/packet_handlers/digging.rs

```diff
@@ -210,41 +193,94 @@ fn handle_shoot_bow(
             player: entity,
         });
     }
+    */
 
-    let arrow_type: Item = match arrow_to_consume {
+    drop(inventory); // Inventory no longer used.
+
+    let _arrow_type: Item = match arrow_to_consume {
         None => Item::Arrow, // Default to generic arrow in creative mode with none in inventory
         Some((_, arrow_stack)) => arrow_stack.ty,
     };
 
-    shoot_arrow_events.single_write(ShootArrowEvent {
-        shooter: Some(entity),
-        position,
-        arrow_type,
-        critical: false, // TODO: Determine critical based on how long bow was pulled back
-    });
+    let timed_use = world.try_get::<ItemTimedUse>(player);
+
+    // Spam clicking can lead to a scenario where this system is called before the UseItem system adds the component
+    // In that case just return.
+    if timed_use.is_none() {
+        return;
+    }
+
+    let timed_use = timed_use.unwrap();
+
+    let mut time_held = game.tick_count - timed_use.tick_start;
+
+    if time_held > 20 {
+        time_held = 20;
+    }
+
+    let charge_force = charge_from_ticks_held(time_held as u32);
+    debug!("Held for {} ticks. Force of {}", time_held, charge_force);
+
+    let init_position = *world.get::<Position>(player) + glm::vec3(0.0, PLAYER_EYE_HEIGHT, 0.0);
+
+    let direction = init_position.direction();
+
+    let arrow_velocity = compute_projectile_velocity(
+        glm::vec3(direction.x, direction.y, direction.z),
+        charge_force as f64,
+        0.0,
+        &mut *game.rng(),
+    );
+    debug!(
+        "Computed exit velocity: {}. Velocity is norm {}",
+        arrow_velocity,
+        arrow_velocity.norm()
+    );
+
+    drop(timed_use);
+
+    world.remove::<ItemTimedUse>(player).unwrap();
+
+    debug!("Preparing to spawn entity");
+
+    let entity = entity::base(init_position)
```

Comment:
This code as well as the `create_spawn_packet` function and the `serialize` function should be probably moved into `entity/arrow.rs` to keep consistent with the item entity implementation.